File: mobile/lib/net/link.dart
/// One open WebSocket, with inbound envelopes buffered into a queue.
///
/// Python gets `await ws.recv()` for free from the websockets library. Dart
/// gives us a single-subscription Stream instead, so we attach exactly one
/// listener for the connection's whole life and park decoded envelopes here.
/// Buffering matters: during the pairing handshake a reply can land before the
/// next `next()` call, and a plain `stream.first` would drop it.
library;

import 'dart:async';

import 'package:web_socket_channel/web_socket_channel.dart';

import 'protocol.dart' as protocol;

class Link {
  Link(this.channel) {
    _sub = channel.stream.listen(
      (raw) {
        final env = protocol.decode(raw);
        if (env != null) _push(env);
      },
      onDone: _fail,
      onError: (_) => _fail(),
      cancelOnError: true,
    );
  }

  final WebSocketChannel channel;
  late final StreamSubscription<dynamic> _sub;
  final _ready = <Map<String, dynamic>>[];
  final _waiting = <Completer<Map<String, dynamic>>>[];
  bool _closed = false;

  /// Fires when the socket drops, for whatever reason.
  final Completer<void> done = Completer<void>();

  void _push(Map<String, dynamic> env) {
    if (_waiting.isNotEmpty) {
      _waiting.removeAt(0).complete(env);
    } else {
      _ready.add(env);
    }
  }

  void _fail() {
    if (_closed) return;
    _closed = true;
    for (final c in _waiting) {
      c.completeError(const SocketDropped());
    }
    _waiting.clear();
    if (!done.isCompleted) done.complete();
  }

  bool get isClosed => _closed;

  /// Next envelope off the wire. Throws [SocketDropped] if the socket closes
  /// first, so callers never hang on a dead link.
  Future<Map<String, dynamic>> next() {
    if (_ready.isNotEmpty) return Future.value(_ready.removeAt(0));
    if (_closed) return Future.error(const SocketDropped());
    final c = Completer<Map<String, dynamic>>();
    _waiting.add(c);
    return c.future;
  }

  void send(Map<String, dynamic> env) {
    if (_closed) return;
    channel.sink.add(protocol.encode(env));
  }

  Future<void> close() async {
    _fail();
    await _sub.cancel();
    await channel.sink.close();
  }
}

class SocketDropped implements Exception {
  const SocketDropped();
  @override
  String toString() => 'socket dropped';
}

File: android/app/src/main/kotlin/com/yoink/net/Link.kt
package com.yoink.net

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.channels.Channel
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * One open WebSocket, with inbound envelopes buffered into a queue.
 *
 * Python gets `await ws.recv()` for free. OkHttp gives us callbacks instead, so
 * frames land in an unlimited [Channel] and [next] reads from it. Buffering
 * matters: during the pairing handshake a reply can arrive before the next
 * `next()` call, and a callback with nowhere to put it would be dropped.
 */
class Link private constructor() {

    private val inbox = Channel<JSONObject>(Channel.UNLIMITED)
    private val opened = CompletableDeferred<Boolean>()
    private var socket: WebSocket? = null

    @Volatile
    var isClosed = false
        private set

    private val listener = object : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            opened.complete(true)
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            Protocol.decode(text)?.let { inbox.trySend(it) }
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) = fail()

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            // Fires for a refused dial too, so unblock anyone awaiting the open.
            opened.complete(false)
            fail()
        }
    }

    private fun fail() {
        if (isClosed) return
        isClosed = true
        // Closing the channel makes a suspended next() throw instead of hanging
        // forever on a socket that is never going to speak again.
        inbox.close()
    }

    /** Next envelope off the wire. Throws once the socket has closed. */
    suspend fun next(): JSONObject = inbox.receive()

    fun send(env: JSONObject) {
        if (!isClosed) socket?.send(Protocol.encode(env))
    }

    fun close() {
        socket?.close(1000, null)
        fail()
    }

    companion object {
        /**
         * Dial [url] and wait for the handshake. Returns null if the peer
         * refused, timed out, or the network is gone.
         */
        suspend fun connect(client: OkHttpClient, url: String): Link? {
            val link = Link()
            link.socket = client.newWebSocket(Request.Builder().url(url).build(), link.listener)
            return if (link.opened.await()) link else { link.close(); null }
        }

        /**
         * pingInterval is OkHttp's own keepalive and is *not* a substitute for
         * the app-level heartbeat in [Peer] — the desktop expects `heartbeat`
         * envelopes and will drop us without them. This just kills a
         * half-open TCP socket faster than the OS would.
         */
        fun newClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .pingInterval(5, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.MILLISECONDS) // a websocket is long-lived
            .build()
    }
}
